Added time range for metrics

maifeeulasad opened the pull request:
closes #1174 

Quick peek:

In long range label are intetionally ommited for clear non-overlapping view:
<img width="1430" height="884" alt="freelens getting metric time range selection feature from maifee ul asad, in long range label are intetionally ommited for clear non-overlapping view" src="https://github.com/user-attachments/assets/7a7e189e-2a98-41a6-8485-1d0c44a780bb" />

In short range more details are visible:
<img width="1430" height="884" alt="freelens getting metric time range selection feature from maifee ul asad, in short range more details are visible" src="https://github.com/user-attachments/assets/ce23cd97-f7f7-4bda-8c25-7995220b3931" />

Lots of predefined time range selection with custom range:
<img width="1430" height="884" alt="freelens getting metric time range selection feature from maifee ul asad, lots of predefined time range selection" src="https://github.com/user-attachments/assets/467961a1-d12d-4268-a229-e5b60faf2701" />

Ability to select custom date along with time, detailed to second:
<img width="1430" height="884" alt="freelens getting metric time range selection feature from maifee ul asad, ability to select custom date along with time detailed to second" src="https://github.com/user-attachments/assets/8eceec8f-4212-4f89-aed3-7904fcdc8be1" />

This is what custom time range looks like, if we click on the date time picker we get even detailed view
<img width="1430" height="884" alt="freelens getting metric time range selection feature from maifee ul asad, this is what custom time range looks like if we click on the date time picker we get even detailed view" src="https://github.com/user-attachments/assets/246929e7-05f9-4a15-8950-482d9e010adf" />

Same view available for pod details as well:
<img width="751" height="896" alt="freelens getting metric time range selection feature from maifee ul asad, same view available for pod details as well" src="https://github.com/user-attachments/assets/0c01d175-b925-433c-9a05-109c9303ba49" />


Open for upcoming reviews! :clinking_glasses:

maifeeulasad committed:
[feat]: metrics time range storage injectable;

maifeeulasad committed:
[feat]: metrics time range selector;

maifeeulasad committed:
[ui]: integrate metrics time range selector in cluster ui;

maifeeulasad committed:
[ui]: integrate metrics time range selector in pod details ui;
 - disabled reload on time range change

maifeeulasad committed:
[feat]: selected metrics time range injectable;

maifeeulasad committed:
[feat]: added time parameter in pod metrics injectable;

maifeeulasad committed:
[feat]: added time parameter in cluster metrics injectable;

maifeeulasad committed:
[feat]: added time parameter in container metrics injectable;

maifeeulasad committed:
[ux]: render bar graph only for selected range; render only necessary labels;

maifeeulasad committed:
[ui]: render only selected time data;
 - cluster
 - container
 - pod details

maifeeulasad committed:
[ux]: enhance switcher layout;
 - flex
 - spacing

maifeeulasad committed:
[lint]: ran all linting command;
 - `pnpm biome:fix`
 - `pnpm prettier:fix`
 - `pnpm trunk:fix`
 - `pnpm lint:fix`

maifeeulasad committed:
Merge branch 'freelensapp:main' into metric-filter-time-range-maifee

maifeeulasad committed:
Merge branch 'main' into metric-filter-time-range-maifee

dex4er committed:
Merge branch 'main' into metric-filter-time-range-maifee

dex4er committed:
Register DI

maifeeulasad committed:
Merge branch 'main' into metric-filter-time-range-maifee

dex4er committed:
Merge branch 'main' into metric-filter-time-range-maifee

MatiasRoje commented:
As already mentioned in the AI review, it would be great if you could adjust these points:
- in packages/core/src/renderer/components/cluster/metrics-time-range-selector.tsx, you are using a plain alert. We have a notification system in Freelens (e.g. using `showErrorNotification`)
- the inline style should be extracted into a module.scss file
 
Other points made by the review are valid as well but those should be addressed before merging.

Other things that I found:
The MetricsTimeRangeSelector is not behaving as a model. That component should be wrapped/behave similarly to components using `<Dialog>` (e.g. in port-forward there is similar code implemented).
And the time picker itself,  the `<input type="datetime-local">`, is not respecting the user theme. I'm using a dark theme but the component is using a light theme. Is that the native chromium input? If our styling doesn't properly work for it, we might need to find an alternative.

Thank you for the great work so far ! @maifeeulasad

maifeeulasad committed:
[refactor]: extract css in dedicated module for `metrics-time-range-selector` form review;

maifeeulasad committed:
[refactor]: extract css in dedicated module for `cluster-metric-switchers` form review;

maifeeulasad committed:
[refactor]: extract css in dedicated module for `pod-details-container-metrics` form review;

maifeeulasad committed:
[refactor]: minor changes from reivew;

maifeeulasad committed:
[lint]: ran all lint;
 - `pnpm biome:fix && pnpm prettier:fix && pnpm trunk:fix && pnpm lint:fix` x2

maifeeulasad committed:
[ux]: added accessibility for `metrics-time-range-selector`;

maifeeulasad committed:
Merge branch 'main' into metric-filter-time-range-maifee

maifeeulasad committed:
[feat]: adaptive step calculation to reduce data points from review;

maifeeulasad committed:
[ux]: rotated x axis label 90 degree for no overlap;

maifeeulasad committed:
[ux]: header for custom time picker; auto focus for time picker; theming for time picker;

maifeeulasad committed:
[fix]: correct mixin usage in `metrics-time-range-selector.module.scss`;

MatiasRoje committed:
improve chart performance, loading behavior, and time range UX

- optimize metrics requests and payloads for large ranges
- align loading behavior across cluster/pod/container charts
- improve long-range chart rendering and time labels (incl. Today)
- update time-range selector UI/labels and layout consistency
- prevent custom-range keyed-instance growth (custom-active)

MatiasRoje commented:
I pushed some changes of my own because multiple small issues were stacking up and causing significant impact in practice, including metrics views taking more than 30 seconds in some cases, plus inconsistent UI behavior across cluster/pod/container flows.

The changes ended up touching several connected areas:
- metrics request and payload performance for larger ranges
- loading behavior consistency across metrics views
- chart and time-range rendering improvements
- selector and layout consistency fixes

Also worth noting: we currently have very limited direct test coverage for these metrics request and loading semantics, and no snapshot coverage for these UI changes.

@maifeeulasad  Thanks for the work on this PR! Please check what I have done and if you feel something should be added or it's not working as expected just change it.

Given how much this PR has grown and how interconnected the changes are, I will also ask another maintainer for an additional review before merge.

MatiasRoje committed:
Merge branch 'main' into metric-filter-time-range-maifee

MatiasRoje committed:
add workloadMemoryUsage to ClusterMetricData and update node metrics handling

MatiasRoje committed:
enhance metrics handling and add time range support for ingresses, nodes, and pvc

MatiasRoje committed:
replace createTimeRangeKey with createMetricsTimeRangeKey and enhance behavior over time and in custom time ranges

MatiasRoje committed:
simplify metrics time range handling, remove cache and unreliable custom fallback

MatiasRoje committed:
add metrics handling for daemonsets, deployments, jobs, replicasets, and statefulsets

maifeeulasad commented:
Hello @MatiasRoje , any reason behind marking this pull request as draft?

MatiasRoje committed:
refactor metrics components and injectables

- remove the MetricsTimeRangeSelector from various metrics detail components, replacing it with TimeRangedResourceMetrics
- updated deployment, job, replica set, stateful set, and pod metrics injectables to utilize createTimeRangedMetricsInjectable for cleaner code and improved maintainability
- general test cleanup and improvements where it matters

MatiasRoje committed:
refactor calculateAdaptiveStep function for improved readability inspired by PR review

MatiasRoje committed:
enhance metrics time range handling and stale behavior

MatiasRoje committed:
add timestamp normalization in requestMetrics

MatiasRoje committed:
reset custom time range values on dialog open

MatiasRoje committed:
restore dynamic time unit selection for BarChart based on time range

MatiasRoje commented:
In my previous review I missed a of couple of critical points, for example that the time range selector was missing in most parts of our codebase (deployments, statefulset, pvc, namespaces, etc. - we really have metrics everywhere). I also took some shortcuts that I was not comfortable with when I reviewed it again last weekend.
UI was already quite good but the implementation lacked testing as I said and added the mentioned problems.

I have been working on improving it so the feature is in its best shape for prod.
@maifeeulasad please review and validate the changes. If something is not clear or not correctly addressed, let's fix it. I might be missing stuff, this feature is quite complex.
I added the time range selected everywhere where we have metrics, added many tests to make sure everything is working as expected and extracted and simplified some behavior so the code is maintainable.

Happy with the current version now, waiting for @dex4er to decide whether this is ready.
PR is up for review again.

maifeeulasad committed:
Merge branch 'main' into metric-filter-time-range-maifee

dex4er committed:
Merge branch 'main' into metric-filter-time-range-maifee

dex4er committed:
Merge branch 'main' into metric-filter-time-range-maifee

dex4er committed:
Merge branch 'main' into metric-filter-time-range-maifee
